Nested tensor subclass support

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127431

When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

tugsbayasgalan committed:
[DRAFT] nested tensor subclass support

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] nested tensor subclass support"


When we have nested tensor subclasses, we need to recurse down to access the underlying real tensor and wrap it in FakeTensor and recursively build back up the nested tensor subclasses. I am not sure if I am passing around the SymbolicContext correctly?


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] nested tensor subclass support"


When we have nested tensor subclasses, we need to recurse down to access the underlying real tensor and wrap it in FakeTensor and recursively build back up the nested tensor subclasses. I am not sure if I am passing around the SymbolicContext correctly?


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -704,6 +704,32 @@ def f(a):
         self.assertEqual(x_ref.grad, x_test.grad)
         self.assertEqual(x_ref_view.grad, x_test_view.grad)
 
+    def test_nested_subclasses(self):
```
I think the E2E test we should try before landing is probably seeing if this PR fixes the Float8 + FSDP issue here: https://github.com/pytorch-labs/float8_experimental/pull/266#issue-2309035931

I was able to repro it by cloning the `float8_experimental` repro, checking out that PR, and running the test command in the summary. Let me know if you run into issues!

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -704,6 +704,32 @@ def f(a):
         self.assertEqual(x_ref.grad, x_test.grad)
         self.assertEqual(x_ref_view.grad, x_test_view.grad)
 
+    def test_nested_subclasses(self):
```
I also think we want more comprehensive testing. A few test cases I can imagine being useful:

(1) function that accepts multiple inputs/outputs, that have different nesting levels (since the bookkeeping is pretty complicated, so we should torture test it a bit). Example: `def f(wrapper(wrapper(tensor)), wrapper(tensor), tensor, wrapper(wrapper(tensor))`

(2) testing that dynamo can handle nested subclasses ok. For example, if one of the inputs to a compiled function is a nested wrapper subclass, dynamo should know to treat a getattr(nested_subclass, inner_tensor) as returning **another** wrapper tensor subclass. And dynamo should handle user code that takes that inner tensor and does things like:

(2a) directly accesses that inner tensor's subclass metadata

(2b) directly accesses that inner tensor's inner tensors

(2c) directly inlines any custom methods defiend on that inner tensor subclass.

I think it might be helpful to add a new wrapper tensor testing subclass with its own custom metadata and custom method(s), something like:
```
class CustomMethodTensor:
    elem: torch.Tensor:
    constant_int = 3
    
    def add_constant(self, other):
        return self + other + self.constant_int
```

(2a) grabs

bdhirsh replied:
Another test that would be good to have is that we get a reasonable error if you have a fw out that is a nested subclass, but at runtime the gradient that comes in is a (non-nested) subclass

tugsbayasgalan committed:
Update on "[DRAFT] nested tensor subclass support"


When we have nested tensor subclasses, we need to recurse down to access the underlying real tensor and wrap it in FakeTensor and recursively build back up the nested tensor subclasses. I am not sure if I am passing around the SymbolicContext correctly?


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -714,13 +729,59 @@ def transform(attr, inner_t):
-            sub = t.type.__tensor_unflatten__(
-                transformed_tensors_dict, t.ctx, outer_size, outer_stride
-            )
+            todo = plain_meta_tensors
```
Any update here?

tugsbayasgalan replied:
Done

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -795,16 +795,30 @@ def empty_create_subclass(
+                    return inner_t
+
+                attr_fqn = prefix + "." + attr if prefix != "" else attr
+                attr_list = attr_fqn.split(".")
```
If all you're going to do to the attr_fqn is split it, why not just pass around a list

tugsbayasgalan committed:
Update on "[DRAFT] nested tensor subclass support"


When we have nested tensor subclasses, we need to recurse down to access the underlying real tensor and wrap it in FakeTensor and recursively build back up the nested tensor subclasses. I am not sure if I am passing around the SymbolicContext correctly?


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] nested tensor subclass support"


When we have nested tensor subclasses, we need to recurse down to access the underlying real tensor and wrap it in FakeTensor and recursively build back up the nested tensor subclasses. I am not sure if I am passing around the SymbolicContext correctly?


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -791,40 +787,61 @@ def empty_create_subclass(
+                        current_context = symbolic_context.inner_contexts[attr]
+
+                    current_source = AttrSource(source, attr)
+                    new_empty_tensor = _empty_create_subclass(
```
You don't have to fix it here, but there's a somewhat prevalent antipattern in this file of doing small recursions on helper functions, rather than calling all the way back to the very top level. I think it should be OK to recurse to the very top call function, and that makes things more general since you can handle composition of things with other things the small helpers don't help. Just calling attention to this.

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -791,40 +787,61 @@ def empty_create_subclass(
-            sub = t.type.__tensor_unflatten__(
-                transformed_tensors_dict, t.ctx, outer_size, outer_stride
+            sub = _empty_create_subclass(
+                t, outer_size, outer_stride, symbolic_context, callback, source
```
ACKing this part, I'll let Brian do the rest

tugsbayasgalan committed:
Update on "[DRAFT] nested tensor subclass support"


When we have nested tensor subclasses, we need to recurse down to access the underlying real tensor and wrap it in FakeTensor and recursively build back up the nested tensor subclasses. I am not sure if I am passing around the SymbolicContext correctly?


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58533224).

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -0,0 +1,69 @@
+
+
+# A simple tensor subclass that holds a tensor with custom metadata and custom method
+class CustomTensor(torch.Tensor):
```
nit: `CustomTensor` is pretty generic-sounding. Since this is mainly used to test custom metadata... `ConstantExtraMetadataTensor` maybe?

bdhirsh commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -0,0 +1,69 @@
+        return return_and_correct_aliasing(func, args, kwargs, out)
+
+
+class CustomTensorMode(torch.utils._python_dispatch.TorchDispatchMode):
```
tbh - mode support with compile isn't great right now, so I would hesitate to add it in our tests (I forget where exactly we needed the TwoTensor mode for testing). Any reason you need it?

bdhirsh commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -0,0 +1,69 @@
+            for o_inner in out_inner_flat
+        ]
+        out = pytree.tree_unflatten(out_flat, spec)
+        return return_and_correct_aliasing(func, args, kwargs, out)
```
fyi given the info about subclass aliasing in this doc, we're probably going to remove this API and not require it for subclasses https://docs.google.com/document/d/1vpbGMsbdhPzFpXi8GJxzkzAGufs0slrIqKqfwDNwKj4/edit#heading=h.76wykr4y5vap

bdhirsh commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -0,0 +1,69 @@
+
+
+# A simple tensor subclass that holds a tensor with custom metadata and custom method
+class CustomTensor(torch.Tensor):
```
actually... is this subclass used anywhere in this PR?

tugsbayasgalan replied:
Yep it is used inside test_aotdispatch.py

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58533224).

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -746,6 +747,109 @@ def f(a):
+        out.sum().backward()
+
+        self.assertTrue(isinstance(custom_aa.grad, ConstantExtraMetadataTensor))
+        self.assertTrue(isinstance(custom_aa.grad.elem, ConstantExtraMetadataTensor))
```
tests that the type of the grad are good, but any reason not to write the test to actually compare results against eager mode?

For this test and all the other ones in this file.

```
f_compiled = torch.compile(f, backend="aot_eager_decomp_partition")

out_ref = f(inp_ref)
out_test = f(inp_test)

out_ref.sum().backward()
out_test.sum().backward()
self.assertEqual(out_Ref, out_test)
self.assertEqual(inp_ref.grad, inp_test.grad)
```

tugsbayasgalan replied:
Done

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1302,7 +1302,7 @@ class SubclassSymbolicContext(StatefulSymbolicContext):
     flexibility, with inner symbolic contexts mapped via attr -> symbolic context.
     """
-    inner_contexts: Dict[str, SymbolicContext] = None
+    inner_contexts: Dict[str, "SubclassSymbolicContext"] = None
```
why the typing change / this is wrong?

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -779,10 +779,6 @@ def empty_create_subclass(
             # NB: t.ctx could be None if the subclass in question has no
             # meaningful context
 
-            assert symbolic_context is None or isinstance(
```
can you explain why we don't need the assert anymore?

bdhirsh replied:
oh nvm, code motion? (I see it lower down)

tugsbayasgalan replied:
Yep i changed this function to be recursive

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58533224).

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -148,15 +148,30 @@ def unset_fake_temporarily():
-        # need to recurse because we could have nested subclasses
-        all_fake = all(is_fake(x) for x in flattened_tensors)
-        any_fake = any(is_fake(x) for x in flattened_tensors)
+        plain_tensors = get_plain_tensors(x)
```
just curious - it seems like the existing version here (prior to your change) was recursive, and already supported nested subclasses? Any reason for the change?

tugsbayasgalan replied:
Yep no idea why i did it this way lol

tugsbayasgalan commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -171,46 +171,51 @@ class SubclassCreationMeta:
-    outer_stride: Tuple[int, ...]
+    original_subclass: Any
+    # If True, then we are constructing metadata at runtime
+    is_runtime: bool
```
Hmm we need this flag because we want to construct SubcassCreationMeta at runtime for the backward part to make sure strides/nested-ness behaviour etc didn't change. Only real time this flag is used is for post_init of SubclassCreationMeta where it asserts the original tensor that is used is FakeTensor which is not true for runtime.

tugsbayasgalan replied:
I think we already do this today? : https://www.internalfb.com/code/fbsource/[70e4cb1526cd6c256299c6eec4c7d40f0367110b]/fbcode/caffe2/torch/_functorch/_aot_autograd/schemas.py?lines=188

bdhirsh replied:
Ah but in that link, we aren't storing an `is_runtime` field on the metadata - the callee is passing it in directly.

Commented more here: https://github.com/pytorch/pytorch/pull/127431/files#r1651720535

tugsbayasgalan commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -171,46 +171,51 @@ class SubclassCreationMeta:
     def __post_init__(self):
         # sanity assert to make sure we don't leak memory
-        assert is_fake(self.original_subclass)
+        if not self.is_runtime:
```
cc @bdhirsh

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1735,10 +1738,32 @@ def backward(ctx, *flat_args):
                 # so I can use aot_dispatch_subclass_wrapper() here.
                 if CompiledFunction.maybe_subclass_metadata is not None:
+                    tangents = all_args[tangents_start_idx:tangents_end_idx]
+                    runtime_subclass_tangent_meta = create_subclass_meta(
```
oh I see. I would prefer to not create fresh `SubclassCreationMeta`s at runtime here:

(1) this is all "hot-path" code, and any python data structures we create at runtime are just extra overhead

(2) AOTAutograd already has a lot of metadata, and knowing that all metadata is statically created at runtime is a nice invariant.

bdhirsh replied:
alternative: the `SubclassCreationMeta` stashes the type of the current subclass as well as its constant metadata, so you can just compare that directly against the runtime subclasses that enter the backward.

bdhirsh commented:
Left a few comments, looks good! Main comment is about the logic to construct metadata at runtime.

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58533224).

bdhirsh commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -171,47 +171,56 @@ class SubclassCreationMeta:
     flat_tensor_start_idx: int
     # The number of tensors that live in this subclass wrapper
     arg_count: int
+    # meta and inner_keys are produced by the subclass's __tensor_flatten__.
```
nit: comment out of date (`inner_keys`)

bdhirsh commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -171,47 +171,56 @@ class SubclassCreationMeta:
+        curr_start_idx = self.flat_tensor_start_idx
+        for attr, creation_meta in self.attrs.items():
+            if creation_meta is None:
+                subclass = all_args[curr_start_idx]
```
nit: the variable name `subclass` here seems misleading, since it may or not actually be a subclass (my understanding is that if `creation_meta` is None, this is guaranteed to be a plain tensor.

Maybe `inner_tensor`?

bdhirsh commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -171,47 +171,56 @@ class SubclassCreationMeta:
     flat_tensor_start_idx: int
     # The number of tensors that live in this subclass wrapper
     arg_count: int
```
after reading the code, some invariants that I think are worth explicitly mentioning in the comments:

- `arg_count` is inclusive of the arg_counts of any inner tensor subclasses: If I have a TwoTensor and both of its inner elements are TwoTensors, then the `arg_count` of the outer-most sublass will be 4

bdhirsh commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -171,47 +171,56 @@ class SubclassCreationMeta:
+                curr_start_idx += creation_meta.arg_count
+            inner_tensors[attr] = subclass
+
+        rebuilt = type(self.original_subclass).__tensor_unflatten__(
```
All of the indices in this reconstruction are definitely non-trivial. It would be great if we had some runtime debug-asserts we could run that would tell us if we messed up the indexing somewhere, so we get a less cryptic error if we get this wrong 🤔. I can't think of a great way to do this though, unless we do something like save all of the shapes of the inner tensors at trace time and assert that our reconstructed inner tensors are have the same shape at runtime

bdhirsh commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -0,0 +1,67 @@
+
+
+# A simple tensor subclass that holds a tensor with custom metadata and custom method
+class ConstantExtraMetadataTensor(torch.Tensor):
```
nit: you renamed the class here, but the file name is still `custom_tensor.py`.

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -746,6 +747,126 @@ def f(a):
+        z = x.clone().detach().requires_grad_()
+        z_compile = z.clone().detach().requires_grad_()
+
+        out_eager = f(x_nested, y_nested, z)
```
hmm... more out of paranoia than anything else, I'm worried about more complicated sets of inputs. The inputs to this test are something like Two(Two(plain, plain), Two(plain, plain)), plain, plain.

Some more testing ideas:
(1) Add a fourth argument that is an unbalanced TwoTensor, e.g. `Two(plain, Two(plain, plain))`
(2) add different subclass types into the test: e.g. make one input `ConstantMetadataTensor(plain)`, and another a `TwoTensor(plain, ConstantMetadataTensor(plain))`.

bdhirsh reviewed (approved):
left some more nits and more tests would be great, pre-emptively stamping!

weifengpy commented:
curious are we landing this PR soon? It's helpful in addressing IMA issues when compiling DTensor(local=fp8). For cpu overhead and gpu time, computing fp8 amax in eager is still faster than torch.compile https://github.com/pytorch/pytorch/issues/129457

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan commented:
> curious are we landing this PR soon? It's helpful in addressing IMA issues when compiling DTensor(local=fp8). Super valuable work!
> 
> sharing my 2 cents perfs. For cpu overhead and gpu time, computing fp8 amax in eager is still faster than torch.compile #129457

Will try to land today :)

tugsbayasgalan committed:
Update on "Nested tensor subclass support"


When we have nested tensor subclasses, we need to recursively flatten/unflatten in Fake tensor creation and AOTAUtograd. Most of the PR is about mechanical change which changes today's single level flatten logic to be recursive. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D58533224](https://our.internmc.facebook.com/intern/diff/D58533224)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58533224).
